### PR TITLE
Alternative config for the Style/AndOr 👮‍♀️

### PR DESCRIPTION
**Pull request by klappradla**
This is more of a suggestion - or a question whether or not this discussion already happened _(if so, feel free to close this)_ - to change the configuration for the [Style/AndOr](http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/AndOr) 👮 

The cop is about replacing `and` and `or` with the logical operators `&&` and `||`.

**However**
Even through `and` and `or` have the same logical meaning as their pure logical counterparts, they are not 100% the same: they have a lower precedence! See [this gist](https://gist.github.com/devboxhl/3d87c36c04f6e628d1e366e11d66e966) for some examples of how this affects Ruby code in the wild🌴.

**👨‍🔬 🔬**
Edward and I ran into a scenario today, where we wanted to return early from a private method in a Rails controller, rendering a certain view. See [the method in the PR](https://github.com/bitcrowd/MyCountryTalks/blob/05fd1fc5fabca405cebb6f894a898e890187b28c/app/controllers/survey_controller.rb#L43) and for more context maybe have a look at [this blog post](https://blog.arkency.com/2014/07/4-ways-to-early-return-from-a-rails-controller/) on _"returning early from a controller"_.

```ruby
def participate
  check_if_open { return }
  @survey_form = SurveyForm.new(partnership: partnership)
end

private

def check_if_open
  render :not_started and yield if event.survey_not_started?
  render :closed and yield if event.survey_closed?
end
```

This particular method only works when using `and`. Logical operators would simply return `true` or `false` 😞 

**tl;dr**
While it definitely makes sense to enforce the usage of `&&` and `||` in comparisons - `and` and `or` may even lead to unhappy surprises there 🙈 - there are actually still usecases where one may deliberately want to use the weaker operators to archive something completely different.
I'd therefore suggest to configure the cop to only enforce the logical operators in conditionals.

Moving from 👇 
```ruby
# bad
foo.save and return

# bad
if foo and bar
end

# good
foo.save && return

# good
if foo && bar
end
```

to 👇 
```ruby
# bad
if foo and bar
end

# good
foo.save && return

# good
foo.save and return

# good
if foo && bar
end
```

### CODE REVIEW

**Comment by tessi** [comment]
As far as I know rubocop assumes that "the average ruby programmer" may not know/understand the differences in precedence over `&& ||` and `and or`. I'm sure we all know here at bitcrowd, but since we often program code for the client (which eventually will maintain the code in the end), I'm *slightly* in favor to still avoid `and or` if possible. This may even lead to a little longer, but more expressive code:

```
render status: 200 and return if user.can_do_stuff?
result = do_something_else
render status: result.to_i

# vs

if user.can_do_stuff
  render status: 200
else
  result = do_something_else
  render status: result.to_i
end
```

🤔 the render-line with `and` totally looks shorter, but it's a lot of code for one line.

**Comment by andreasknoepfle** [comment]
I'm not completely sure what the problem with `and` and `&&` differences here is in this particular case. I only know, that the rendered value does not depend on the return value of the method, so I'm pretty sure one should be able to do 
`render :new && return` 

I guess the reason why the above does not work is probably more likely in the use of the block with `{ return }` 
I'm not entirely sure about this, but in this case I think a `before_action` would be an exact fit, since a `before_action` automatically returns if it renders or redirects. Usually @tessi has strong arguments against `before_action`s due to readability and explicitness, but I'd say in this case the `check_if_open { return }` is weird enough to be less readable :/

**Comment by klappradla** [comment]
Thanx for your explanations @andreasknoepfle 💚 

Turns out my scenario probably is sort of an edge case anyways 😉 

I agree, the `before_action` would be super neat here and it also works with just calling `render` once and having it return ✅ . However, I'm also using an `after_action` filter to remove some header in this controller - and this one will not be executed when returning from a `before_action` callback 😢 

Your other approach
```ruby
render :thing && return
```
Does not fail due to the block statement, but because it's a separate method. The layout looks like this:
```ruby
def endpoint
  check_condition_and_return_early
  # do the actual method stuff
end

private

def check_condition_and_return_early
  render :thing and return if condition?
  render :other_thing and return if other_condition?
end
```

☝️ this works but using `&&` will just execute the original action.

The reason I used the block syntax was just to make it make the `endpoint` method specify the whole "flow" - so that the actual `return` is part of this method instead of just somewhere hidden down in the private method.

Anyways, I think there are enough ways around not disabling this cop, so that I can savely close this PR. Thanx for you comments 🙏
